File: backend/GameApp.Application/Services/SceneServices/SceneUpdateService.cs
using GameApp.Application.UseCases.SceneUseCases;
using GameApp.Domain.Entities;
using GameApp.Domain.Repositories;

namespace GameApp.Application.Services.SceneServices;

public class SceneUpdateService : SceneUpdateUseCase
{
    private readonly ISceneRepository _repo;

    public SceneUpdateService(ISceneRepository repo) => _repo = repo;

    public async Task<Scene?> UpdateSceneAsync(Guid id, Scene scene)
    {
        var existingScene = await _repo.FetchByIdAsync(id);
        if (existingScene is null)
        {
            Console.WriteLine($"Scene to update with id {id} not found.");
            return null;
        }

        // If the scene name is going to change
        if (!existingScene.GetName().Equals(scene.GetName()))
        {
            var sceneWithSameName = await _repo.FetchByName(scene.GetName());
            if (sceneWithSameName is not null)  // scene name must be unique in the collection
            {
                Console.WriteLine($"Scene with name '{scene.GetName()}' already exists. Cannot update scene");
                return null;
            }
        }

        return await _repo.UpdateAsync(id, scene);
    }
}

File: frontend/gameapp-frontend/src/application/services/scene-services/scene-update-service.ts
import type { ISceneUpdateUseCase } from "../../usecases/scene-use-cases/scene-update-use-case";
import type { ISceneRepository } from "../../repositories/iscene-repository";
import { Scene } from "../../../domain/entities/scenes/scene";

export class SceneUpdateService implements ISceneUpdateUseCase {
    constructor(private sceneRepository: ISceneRepository) {}

    async updateScene(id: string, scene: Scene): Promise<Scene> {
        const updatedScene = await this.sceneRepository.update(id, scene);
        return updatedScene;
    }
}
